src/internet_identity/src/storage.rs

```diff
@@ -56,69 +87,115 @@
 use ic_stable_structures::Memory;
-use internet_identity_interface::UserNumber;
+use internet_identity_interface::{MigrationState, UserNumber};
 use std::convert::TryInto;
 use std::fmt;
 use std::io::{Read, Write};
-use std::marker::PhantomData;
 use std::ops::RangeInclusive;
 
 #[cfg(test)]
 mod tests;
 
 // version 0: invalid
 // version 1: genesis layout, might have persistent state
-// version 2: genesis layout, must have persistent state
-// version 3+: invalid
-const SUPPORTED_LAYOUT_VERSIONS: RangeInclusive<u8> = 1..=2;
+// version 2: migration in progress genesis -> post-migration layout, must have persistent state
+// version 3: post-migration layout
+const SUPPORTED_LAYOUT_VERSIONS: RangeInclusive<u8> = 1..=3;
 
 /// Reserved space for the header before the anchor records start.
-const RESERVED_HEADER_BYTES: u64 = 512;
-const DEFAULT_ENTRY_SIZE: u16 = 2048;
+const WASM_PAGE_SIZE: u64 = 65_536;
+
+const RESERVED_HEADER_BYTES_V1: u64 = 512;
+const RESERVED_HEADER_BYTES_V3: u64 = 2 * WASM_PAGE_SIZE; // 1 page reserved for II config, 1 for memory manager
+
+const DEFAULT_ENTRY_SIZE_V1: u16 = 2048;
+const DEFAULT_ENTRY_SIZE_V3: u16 = 4096;
+
 const EMPTY_SALT: [u8; 32] = [0; 32];
 const GB: u64 = 1 << 30;
-const STABLE_MEMORY_SIZE: u64 = 8 * GB;
+
+/// In practice, II has 32 GB of stable memory available. But we want to keep the default
+/// user range until the stable memory migration is complete. Thus we keep this value for anchor
+/// range checking for the time being.
+const LEGACY_STABLE_MEMORY_SIZE: u64 = 8 * GB;
 /// We reserve last ~10% of the stable memory for later new features.
-const STABLE_MEMORY_RESERVE: u64 = STABLE_MEMORY_SIZE / 10;
+const STABLE_MEMORY_RESERVE: u64 = LEGACY_STABLE_MEMORY_SIZE / 10;
 
 const PERSISTENT_STATE_MAGIC: [u8; 4] = *b"IIPS"; // II Persistent State
 
 /// The maximum number of users this canister can store.
-pub const DEFAULT_RANGE_SIZE: u64 =
-    (STABLE_MEMORY_SIZE - RESERVED_HEADER_BYTES - STABLE_MEMORY_RESERVE)
-        / DEFAULT_ENTRY_SIZE as u64;
+pub const DEFAULT_RANGE_SIZE_V1: u64 =
+    (LEGACY_STABLE_MEMORY_SIZE - RESERVED_HEADER_BYTES_V1 - STABLE_MEMORY_RESERVE)
+        / DEFAULT_ENTRY_SIZE_V1 as u64;
 
 pub type Salt = [u8; 32];
 
 /// Data type responsible for managing user data in stable memory.
-pub struct Storage<T, M> {
+pub struct Storage<M> {
     header: Header,
     memory: M,
-    _marker: PhantomData<T>,
 }
 
 #[repr(packed)]
 struct Header {
     magic: [u8; 3],
     // version 0: invalid
     // version 1: genesis layout, might have persistent state
-    // version 2: genesis layout, must have persistent state
-    // version 3+: invalid
+    // version 2: migration in progress genesis -> post-migration layout, must have persistent state
+    // version 3: post-migration layout
     version: u8,
     num_users: u32,
     id_range_lo: u64,
     id_range_hi: u64,
     entry_size: u16,
     salt: [u8; 32],
+    entry_size_new: u16,       // post-migration entry size
```

Comment:
I'm actually a little confused (not 100% related to this PR): why do we store the entry size in the header?

Reply:
It was this way before. But I would assume for the following reasons:
* storage used to be generic, so it made sense to store the entry size in the data structure itself
* reading the data from stable memory makes sure that it matches the rest of the data structure. I.e. reading it from a constant could lead to bugs where the constant changes without migrating the data.